src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CancelMessage.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.cloud.spanner.pgadapter.wireprotocol;
+
+import com.google.cloud.spanner.pgadapter.ConnectionHandler;
+import java.text.MessageFormat;
+
+/**
+ * This message handles the imperative cancellation, as issues in a new connection by the PG wire
+ * protocol. We expect that this message contains an ID for the connection which issues the
+ * original query, as well as an auth secret (which we ignore, as all connections are localhost).
```

Comment:
I think that may be the case, however, since we only accept localhost connections, only local applications can issue the cancel message to the proxy, so, the only denial of service would come from the host itself, and given full access to the host, it would be just as easy to sniff loopback and gain access to the password (especially since the local connection is not encrypted).

Reply:
Done :)